src/Access/AccessControl.cpp

```diff
@@ -72,18 +72,26 @@ class AccessControl::ContextAccessCache
 
     std::shared_ptr<const ContextAccess> getContextAccess(const ContextAccessParams & params)
     {
-        std::lock_guard lock{mutex};
-        auto x = cache.get(params);
-        if (x)
         {
-            if ((*x)->tryGetUser())
-                return *x;
-            /// No user, probably the user has been dropped while it was in the cache.
-            cache.remove(params);
+            std::lock_guard lock{mutex};
+            auto x = cache.get(params);
+            if (x)
+            {
+                if ((*x)->getUserID() && !(*x)->tryGetUser())
+                    cache.remove(params); /// The user has been dropped while it was in the cache.
+                else
+                    return *x;
+            }
         }
+
         auto res = std::make_shared<ContextAccess>(access_control, params);
         res->initialize();
-        cache.add(params, res);
+
+        {
+            std::lock_guard lock{mutex};
+            cache.add(params, res);
```

Comment:
is it okay that we possible could recompute and rewrite the same cache entry concurrently (leaving the first returned result removed from cache)?

Reply:
Yes, it can recompute the same cache entry concurrently. It doesn't seem like a big problem here but I'll try to fix that anyway.

Reply:
No, it appears it's not quite easy to do that properly. This my change is not 100%-related to the main problem I'm fixing here, so I think I'll revert my changes here, write a comment, and leave the rest for a separate PR.